File: examples/rust-app-demo/src/bin/seed_demo.rs
use awa::{insert_with, InsertOpts, JobState};
use serde_json::json;
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tracing_subscriber::EnvFilter;
use uuid::Uuid;

#[path = "../shared.rs"]
mod shared;

use shared::{
    build_demo_client, clear_demo_data, create_checkout, create_pool, hero_scheduled_time,
    prepare_schema, seed_available_cache_jobs, seed_failed_syncs, seed_pending_payments,
    seed_scale, seed_scheduled_reports, wait_for_cron_sync, wait_for_many, CACHE_QUEUE, CRON_NAME,
    EMAIL_QUEUE, GenerateRevenueReport, OPS_QUEUE, PAYMENTS_QUEUE, REPORTS_QUEUE,
    WarmProductCache,
};

fn parse_scale() -> &'static str {
    let mut args = std::env::args().skip(1);
    while let Some(arg) = args.next() {
        if arg == "--scale" {
            return match args.next().as_deref() {
                Some("small") => "small",
                Some("large") => "large",
                Some("video") => "video",
                _ => "medium",
            };
        }
    }
    "medium"
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt()
        .with_env_filter(EnvFilter::from_default_env())
        .init();

    let scale_name = parse_scale();
    let preset = seed_scale(scale_name);

    let pool = create_pool().await?;
    prepare_schema(&pool).await?;
    clear_demo_data(&pool).await?;

    let callback_ids = Arc::new(Mutex::new(Vec::<Uuid>::new()));
    let client = build_demo_client(pool.clone(), Some(callback_ids.clone()))?;

    let mut receipt_job_ids = Vec::with_capacity(preset.completed_orders);
    for i in 0..preset.completed_orders {
        let checkout = create_checkout(
            &pool,
            &format!("customer{}@example.com", i + 1),
            (1499 + (i * 200)) as i32,
            Some(format!("ord_demo_{:03}", i + 1)),
        )
        .await?;
        if let Some(job_id) = checkout.confirmation_job_id {
            receipt_job_ids.push(job_id);
        }
    }

    let failed_ids = seed_failed_syncs(&pool, preset.failed_syncs).await?;
    let waiting_ids = seed_pending_payments(&pool, preset.waiting_payments).await?;
    seed_available_cache_jobs(&pool, preset.available_cache_jobs).await?;
    seed_scheduled_reports(&pool, preset.scheduled_reports).await?;

    let hero_available = insert_with(
        &pool,
        &WarmProductCache { slug: "/".into() },
        InsertOpts {
            queue: CACHE_QUEUE.to_string(),
            tags: vec!["demo".into(), "hero".into(), "cache".into()],
            metadata: json!({"app": "rust-demo-shop", "hero": true}),
            ..Default::default()
        },
    )
    .await?;

    let hero_scheduled = insert_with(
        &pool,
        &GenerateRevenueReport {
            report_name: "weekly-gross-margin".into(),
        },
        InsertOpts {
            queue: REPORTS_QUEUE.to_string(),
            run_at: Some(hero_scheduled_time()),
            tags: vec!["demo".into(), "hero".into(), "reports".into()],
            metadata: json!({"app": "rust-demo-shop", "hero": true}),
            ..Default::default()
        },
    )
    .await?;

    // For "video" scale, don't start workers or wait for state transitions.
    // Jobs stay in their initial states (available, scheduled) so the user
    // can start workers separately and watch progress live in the UI.
    let callback_id = if scale_name == "video" {
        // Still need to briefly start the client to sync the cron schedule,
        // but don't process any jobs.
        client.start().await?;
        wait_for_cron_sync(&pool, Duration::from_secs(10)).await?;
        client.shutdown(Duration::from_secs(2)).await;
        None
    } else {
        client.start().await?;

        wait_for_many(
            &pool,
            &receipt_job_ids,
            JobState::Completed,
            Duration::from_secs(12),
        )
        .await?;
        wait_for_many(
            &pool,
            &failed_ids,
            JobState::Failed,
            Duration::from_secs(60),
        )
        .await?;
        wait_for_many(
            &pool,
            &waiting_ids,
            JobState::WaitingExternal,
            Duration::from_secs(30),
        )
        .await?;
        wait_for_cron_sync(&pool, Duration::from_secs(10)).await?;

        client.shutdown(Duration::from_secs(5)).await;

        callback_ids
            .lock()
            .expect("callback id lock")
            .first()
            .copied()
    };

    println!("\nRust demo data ready");
    println!("====================");
    println!("scale preset:       {scale_name}");
    println!(
        "queues:             {EMAIL_QUEUE}, {PAYMENTS_QUEUE}, {OPS_QUEUE}, {CACHE_QUEUE}, {REPORTS_QUEUE}"
    );
    println!(
        "counts:             completed={} failed={} waiting={} available={} scheduled={}",
        receipt_job_ids.len(),
        failed_ids.len(),
        waiting_ids.len(),
        preset.available_cache_jobs + 1,
        preset.scheduled_reports + 1
    );
    println!(
        "hero available:     id={} state={:?}",
        hero_available.id, hero_available.state
    );
    println!(
        "hero scheduled:     id={} state={:?}",
        hero_scheduled.id, hero_scheduled.state
    );
    println!(
        "callback id:        {}",
        callback_id
            .map(|id| id.to_string())
            .unwrap_or_else(|| "missing".into())
    );
    println!("cron job:           {CRON_NAME}");
    println!("\nSuggested next steps:");
    println!("  cargo run -p awa-cli -- --database-url $DATABASE_URL serve");
    println!("  cargo run --bin app");
    println!("  cargo run --bin workers");

    Ok(())
}

File: examples/python-app-demo/seed_demo.py
"""Seed the demo app and Awa UI with realistic domain data."""

import argparse
import asyncio

import awa

from demo_app.shared import (
    CACHE_QUEUE,
    CRON_NAME,
    EMAIL_QUEUE,
    GenerateRevenueReport,
    OPS_QUEUE,
    PAYMENTS_QUEUE,
    REPORTS_QUEUE,
    SEED_SCALE_PRESETS,
    WarmProductCache,
    clear_demo_data,
    create_checkout,
    create_client,
    ensure_app_schema,
    register_workers,
    seed_available_cache_jobs,
    seed_failed_syncs,
    seed_pending_payments,
    seed_scheduled_reports,
)


async def wait_for_state(
    client: awa.AsyncClient,
    job_id: int,
    expected_state: awa.JobState,
    *,
    timeout_seconds: float = 12.0,
) -> awa.Job:
    deadline = asyncio.get_running_loop().time() + timeout_seconds
    while True:
        job = await client.get_job(job_id)
        if job.state == expected_state:
            return job
        if asyncio.get_running_loop().time() >= deadline:
            raise TimeoutError(
                f"job {job_id} never reached {expected_state}; last state was {job.state}"
            )
        await asyncio.sleep(0.1)


async def wait_for_many(
    client: awa.AsyncClient,
    job_ids: list[int],
    expected_state: awa.JobState,
    *,
    timeout_seconds: float = 12.0,
) -> None:
    for job_id in job_ids:
        await wait_for_state(
            client,
            job_id,
            expected_state,
            timeout_seconds=timeout_seconds,
        )


async def wait_for_cron_sync(client: awa.AsyncClient, name: str, *, timeout_seconds: float = 10.0) -> None:
    deadline = asyncio.get_running_loop().time() + timeout_seconds
    while True:
        tx = await client.transaction()
        row = await tx.fetch_optional(
            "SELECT 1 FROM awa.cron_jobs WHERE name = $1",
            name,
        )
        await tx.commit()
        if row is not None:
            return
        if asyncio.get_running_loop().time() >= deadline:
            raise TimeoutError(f"cron job {name} was not synced")
        await asyncio.sleep(0.1)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Seed the demo FastAPI app and Awa UI.")
    parser.add_argument(
        "--scale",
        choices=sorted(SEED_SCALE_PRESETS.keys()),
        default="medium",
        help="Dataset size preset. Use 'medium' for video and 'large' for a denser dashboard.",
    )
    return parser.parse_args()


async def main() -> None:
    args = parse_args()
    preset = SEED_SCALE_PRESETS[args.scale]

    client = create_client()
    await client.migrate()
    await ensure_app_schema(client)
    await clear_demo_data(client)

    callback_ids: list[str] = []
    register_workers(client, callback_ids)

    receipt_job_ids: list[int] = []
    for i in range(preset["completed_orders"]):
        checkout = await create_checkout(
            client,
            customer_email=f"customer{i + 1}@example.com",
            total_cents=1499 + (i * 200),
            order_id=f"ord_demo_{i + 1:03d}",
        )
        receipt_job_ids.append(int(checkout["confirmation_job_id"]))

    failed_ids = await seed_failed_syncs(client, preset["failed_syncs"])
    waiting_ids = await seed_pending_payments(client, preset["waiting_payments"])
    await seed_available_cache_jobs(client, preset["available_cache_jobs"])
    await seed_scheduled_reports(client, preset["scheduled_reports"])

    hero_available = await client.insert(
        WarmProductCache(slug="/"),
        queue=CACHE_QUEUE,
        tags=["demo", "hero", "cache"],
        metadata={"app": "demo-shop", "hero": True},
    )
    hero_scheduled = await client.insert(
        GenerateRevenueReport(report_name="weekly-gross-margin"),
        queue=REPORTS_QUEUE,
        run_at="2099-01-01T00:00:00+00:00",
        tags=["demo", "hero", "reports"],
        metadata={"app": "demo-shop", "hero": True},
    )

    await client.start(
        [(EMAIL_QUEUE, 2), (OPS_QUEUE, 1), (PAYMENTS_QUEUE, 1)],
        leader_election_interval_ms=100,
        heartbeat_interval_ms=100,
        promote_interval_ms=100,
    )

    await wait_for_many(client, receipt_job_ids, awa.JobState.Completed)
    await wait_for_many(client, failed_ids, awa.JobState.Failed)
    await wait_for_many(client, waiting_ids, awa.JobState.WaitingExternal)
    await wait_for_cron_sync(client, CRON_NAME)
    await client.shutdown()

    print("\nDemo app data ready")
    print("===================")
    print(f"scale preset:       {args.scale}")
    print(
        "queues:             "
        f"{EMAIL_QUEUE}, {PAYMENTS_QUEUE}, {OPS_QUEUE}, {CACHE_QUEUE}, {REPORTS_QUEUE}"
    )
    print(
        "counts:             "
        f"completed={len(receipt_job_ids)} "
        f"failed={len(failed_ids)} "
        f"waiting={len(waiting_ids)} "
        f"available={preset['available_cache_jobs'] + 1} "
        f"scheduled={preset['scheduled_reports'] + 1}"
    )
    print(f"hero available:     id={hero_available.id} state={hero_available.state}")
    print(f"hero scheduled:     id={hero_scheduled.id} state={hero_scheduled.state}")
    print(f"callback id:        {callback_ids[0] if callback_ids else 'missing'}")
    print(f"cron job:           {CRON_NAME}")
    print("\nSuggested next steps:")
    print("  (cd ../../awa-ui/frontend && npm install && npm run build)  # once per checkout")
    print("  uv run awa --database-url $DATABASE_URL serve")
    print("  uv run uvicorn demo_app.app:app --reload")


if __name__ == "__main__":
    asyncio.run(main())
